### DISCUSSION TITLE
Allow method of setting `user` or `authContext` on EntityManager

### DISCUSSION THREAD

**Discussion by kpervin** [mikro-orm/mikro-orm#5219] [Ideas]
Given that MikroORM already utilizes AsyncLocalStorage for the RequestContext, I am wondering if it would be at all possible to add on a means to append an auth context to a currently running EntityManager. This would allow things like Passport or other middleware to then set the current user context per request lifecycle so that MikroORM would then be able to utilize this "user" within things such as Event Lifecycle Hooks, Filters, or even potentially parsing default values.

The actual setting of the current EntityManager user context would have to be done by the user in some form of middleware/guard/interceptor/etc, but allowing it to be accessible across the EntityManager lifecycle would be a huge plus.

So something along the lines of this:

```ts
export async function bootstrap(port = 3001) {
  const orm = await MikroORM.init();
  const app = fastify();

  // register request context hook
  app.addHook('onRequest', (request, reply, done) => {
    orm.em.setUser(request.user);
    RequestContext.create(orm.em, done);
  });

  // shut down the connection when closing the app
  app.addHook('onClose', async () => {
    await orm.close();
  });

  // register routes here
  // ...

  const url = await app.listen({ port });

  return { app, url };
}

...

class Foo {
  @PrimaryKey()
  id: number;

  @Property()
  createdAt = new Date();
  
  @ManyToOne()
  createdBy: User;
}

class Subscriber implements EventSubscriber<Foo> {
  beforeCreate(args: EventArgs<Foo>) {
    args.changeSet.payload.createdBy = args.em.getUser();
  }
}
```

 Curious to your thoughts @B4nan.

**Comment by B4nan** [comment]
You can already use the filter parameters for this, or the new logger context, both work similarly - as a generic key-value storage scoped to the fork, but if you set it on the global EM, it will be copied to the fork. Both have a getter and setter on the EM class:

https://mikro-orm.io/api/core/class/EntityManager#getFilterParams
https://mikro-orm.io/api/core/class/EntityManager#setFilterParams
https://mikro-orm.io/api/core/class/EntityManager#getLoggerContext
https://mikro-orm.io/api/core/class/EntityManager#setLoggerContext

**Reply by kpervin** [reply]
Which of those would be available in Lifecycle hooks? That's really the place I'm most looking to use something like this, and I'm sure others would also find it handy.

I would also argue that having something dedicated would be worthwhile, even if it's not a dedicated `user` scenario. Maybe something along the lines of a `custom_metadata` field on the EntityManager that could transport data across the lifecycle of the RequestContext.

**Reply by B4nan** [reply]
Both of them, as you get the EM instance in there.

**Reply by kpervin** [reply]
In all honesty, those two methods aren't exactly straight forward or self-apparent to a dev, as well as not being able to find anything about logger context in the docs. 

My proposal would be to add some sort of object/store over request lifecycle such as `executiontContextMetadata` or something along those lines that could have any number of key/value pairs set during the execution context lifecycle which would be obtainable when having access to the EntityManager. It would also be worthwhile to be able to add types to it in some capacity, either through namespacing, setting a type directly on the `get` method, or maybe adding something to the config file (unsure how that would work without namespacing).

As you've mentioned, there are technically already ways of doing this, but they're not self-apparent for this use case. If `loggerContext` already fits this bill, I would propose renaming it to `executionContextMetadata` or something similar and documenting it.

What are your thoughts @B4nan?

**Reply by B4nan** [reply]
>  as well as not being able to find anything about logger context in the docs.

Some docs are here, but they don't talk about the `get/setLoggerContext()` helper on EM level:

https://mikro-orm.io/docs/logging#providing-additional-context-to-a-custom-logger

>  If loggerContext already fits this bill, I would propose renaming it to executionContextMetadata or something similar and documenting it.

That would have to wait for next major version. But yes, it is exactly the same thing as you want, it is a generic KV storage on the EM instance (so a fork created for a given request), that will be also passed to the logger.

Maybe we could have an alias to it as a get/set pair.

**Reply by kpervin** [reply]
> Maybe we could have an alias to it as a get/set pair.

I can put in a PR for this if you'd like?

**Reply by B4nan** [reply]
Sure

**Reply by kpervin** [reply]
Should we also add in some docs regarding it?

**Reply by B4nan** [reply]
Yes
